cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -121,6 +122,137 @@ Result<std::unique_ptr<KernelState>> CountInit(KernelContext*,
+}
+
+template <typename Type, typename VisitorArgType = typename Type::c_type>
+struct CountDistinctKernel {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    using PhysicalType = typename Type::PhysicalType;
+    AddCountDistinctKernel<PhysicalType, VisitorArgType>(type, func);
+  }
+};
+
+template <>
+struct CountDistinctKernel<FixedSizeBinaryType, util::string_view> {
+  static void Add(InputType type, ScalarAggregateFunction* func) {
+    AddCountDistinctKernel<FixedSizeBinaryType, util::string_view>(type, func);
+  }
+};
+
+void AddCountDistinctKernels(ScalarAggregateFunction* func) {
+  // Boolean
+  aggregate::CountDistinctKernel<BooleanType>::Add(boolean(), func);
+  // Number
+  aggregate::CountDistinctKernel<Int8Type>::Add(int8(), func);
+  aggregate::CountDistinctKernel<Int16Type>::Add(int16(), func);
+  aggregate::CountDistinctKernel<Int32Type>::Add(int32(), func);
+  aggregate::CountDistinctKernel<Int64Type>::Add(int64(), func);
+  aggregate::CountDistinctKernel<UInt8Type>::Add(uint8(), func);
+  aggregate::CountDistinctKernel<UInt16Type>::Add(uint16(), func);
+  aggregate::CountDistinctKernel<UInt32Type>::Add(uint32(), func);
+  aggregate::CountDistinctKernel<UInt64Type>::Add(uint64(), func);
+  aggregate::CountDistinctKernel<HalfFloatType>::Add(float16(), func);
+  aggregate::CountDistinctKernel<FloatType>::Add(float32(), func);
+  aggregate::CountDistinctKernel<DoubleType>::Add(float64(), func);
+  // Date
+  aggregate::CountDistinctKernel<Date32Type>::Add(date32(), func);
+  aggregate::CountDistinctKernel<Date64Type>::Add(date64(), func);
+  // Time
+  aggregate::CountDistinctKernel<Time32Type>::Add(match::SameTypeId(Type::TIME32), func);
+  aggregate::CountDistinctKernel<Time64Type>::Add(match::SameTypeId(Type::TIME64), func);
+  // Timestamp & Duration
+  aggregate::CountDistinctKernel<TimestampType>::Add(match::SameTypeId(Type::TIMESTAMP),
+                                                     func);
+  aggregate::CountDistinctKernel<DurationType>::Add(match::SameTypeId(Type::DURATION),
+                                                    func);
+  // Interval
+  aggregate::CountDistinctKernel<MonthIntervalType>::Add(month_interval(), func);
+  aggregate::CountDistinctKernel<DayTimeIntervalType>::Add(day_time_interval(), func);
+  aggregate::CountDistinctKernel<MonthDayNanoIntervalType>::Add(month_day_nano_interval(),
+                                                                func);
+  // Binary & String
+  aggregate::CountDistinctKernel<BinaryType, util::string_view>::Add(match::BinaryLike(),
+                                                                     func);
+  aggregate::CountDistinctKernel<LargeBinaryType, util::string_view>::Add(
+      match::LargeBinaryLike(), func);
+  // Fixed binary & Decimal
+  aggregate::CountDistinctKernel<FixedSizeBinaryType, util::string_view>::Add(
+      match::FixedSizeBinaryLike(), func);
```

Comment:
It would be nice to avoid enumerating types explicitly. Can you take a look at the helpers exposed by `codegen_internal.h`, for example `GeneratePhysicalNumeric`, `GenerateTypeAgnosticVarBinaryBase`. One might also add a `GenerateTypeAgnosticFixedBinaryBase`.

Reply:
Thanks, but I as mentioned before: I already look at those helpers and it seems those are for generating the exec part of the kernel. Do you have anything specific in mind? If so, maybe this can be a follow-up for a nice to have/code style.

Reply:
Ah, I hadn't noticed these comments. But then a similar approach to `ProductInit` can be used.

Reply:
Sure it can, but my point in that thread was that in the future we will need to refactor this anyways (to support more complex types). So my suggestion was to leave this PR as it is and try to create follow-ups tickets for these internal implementation details (I think @edponce was doing that for other PRs too)

Reply:
Well, as you prefer, but I don't understand why it would be better to do something later when you could do it now...

Reply:
Thanks! My understanding, from your previous comments, was that we don't want to tackle the nested types in this PR yet (i.e. that we don't want to refactor the code for aggregations now)
Perhaps it was a misunderstanding, let me know otherwise please.

Reply:
"refactor" and "implement nested types" are two different concerns, aren't they? That said, I'm ok with deferring the refactor to another JIRA/PR if you plan to work on it :-)